framework/audio/internal/midiplayer.cpp

```diff
@@ -239,131 +254,81 @@ bool Sequencer::sendEvents(tick_t fromTick, tick_t toTick)
-        if (chState.muted || SKIP_EVENTS.find(event.type()) != SKIP_EVENTS.end()) {
-            // noop
-        } else {
+        if (event && !chState.muted) {
             auto s = synth(event.channel());
             s->handleEvent(event);
             s->setIsActive(true);
+
+            if (event.isChannelVoice() && event.opcode() == midi::Event::Opcode::NoteOn) {
+                auto noteOff = event;
+                noteOff.setOpcode(midi::Event::Opcode::NoteOff);
+                m_noteCache[event.note()] = noteOff;
+            } else if (event.isChannelVoice() && event.opcode() == midi::Event::Opcode::NoteOff) {
+                m_noteCache[event.note()] = Event::NOOP();
+            }
         }
 
         ++pos;
     }
 
+    midiPortDataSender()->sendEvents(fromTick, toTick);
     return true;
 }
 
-float Sequencer::getAudio(float sec, float* buf, unsigned int samples, Context* ctx)
+void MIDIPlayer::sendClear()
 {
-    process(sec, ctx);
-
-    unsigned int totalSamples = samples * AUDIO_CHANNELS;
-
-    // write buffers
-    for (SynthState& state : m_synthStates) {
-        if (state.synth->isActive()) {
-            if (state.buf.size() < totalSamples) {
-                state.buf.resize(totalSamples);
-            }
-            std::memset(&state.buf[0], 0, totalSamples * sizeof(float));
-            state.synth->writeBuf(&state.buf[0], samples);
-        }
-    }
-
-    // mix
-    std::memset(buf, 0, totalSamples * sizeof(float));
-    for (SynthState& state : m_synthStates) {
-        if (state.synth->isActive()) {
-            for (unsigned int s = 0; s < totalSamples; ++s) {
-                buf[s] += state.buf[s];
-            }
+    for (auto& cache: m_noteCache) {
+        auto event = cache.second;
+        if (event) {
+            auto s = synth(event.channel());
+            s->handleEvent(event);
+            midiPortDataSender()->sendSingleEvent(event);
         }
     }
-
-    float cur_sec = static_cast<float>(m_curMSec) / 1000.f;
-    return cur_sec;
+    m_noteCache.clear();
 }
 
-bool Sequencer::hasEnded() const
+void MIDIPlayer::play()
 {
-    if (m_midiStream->isStreamingAllowed && m_streamState.requested) {
-        return false;
-    }
-
-    tick_t prev = ticks(m_prevMSec);
-    if (prev >= m_midiStream->lastTick) {
-        return true;
+    if (m_midiStream && status() != Status::Error) {
+        setStatus(Status::Running);
     }
-
-    return false;
 }
 
-Sequencer::Status Sequencer::status() const
+void MIDIPlayer::stop()
 {
-    return m_status;
-}
-
-bool Sequencer::run(float init_sec)
-{
-    if (m_status == Running) {
-        return true;
+    if (status() != Status::Error) {
+        setStatus(Status::Stoped);
     }
-
-    if (m_status == Error) {
-        return false;
-    }
-
-    m_prevMSec = static_cast<uint64_t>(init_sec * 1000);
-    m_curMSec = m_prevMSec;
-
-    m_status = Running;
-
-    return true;
+    sendClear();

```

Comment:
This is not thread-safe code, it may crash.
This method is called from the main thread, from `void PlaybackController::stop()`

But `m_noteCache` changes from the audio worker thread.
It may be that writing to `m_noteCache` in the audio worker thread will occur simultaneously with clearing `m_noteCache` in the main thread. As result will crash.

Reply:
fixed